### PR TITLE
[DEV-1948] add additionalFee + fee calculation refactoring

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

**Comment by Yannick1712** [comment]
Add migration

File: src/subdomains/supporting/payment/dto/fee.dto.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+export class FeeDto {
+  percentAmount: number;
+  additionalAmount: number;
+}
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/dto/fee.dto.ts:4]: Naming can be improved

File: src/subdomains/supporting/payment/services/fee.service.ts [hunk 2]
```diff
@@ -119,36 +122,57 @@ export class FeeService {
...
-    const negativeDiscountFee = Math.min(...fees.filter((f) => f.type === FeeType.DISCOUNT).map((f) => f.value), 0);
-    const discountFee = positiveDiscountFee + negativeDiscountFee;
-
-    if (baseFee === Infinity) throw new InternalServerErrorException('Base fee is missing');
-    if (baseFee - discountFee < 0) {
+  private calculateFee(fees: Fee[], userDataId?: number): FeeDto {
+    // filter customFee with min. value
+    const customFee = fees
+      .filter((fee) => fee.type === FeeType.CUSTOM)
+      .reduce((minFee, fee) => (fee.value > minFee?.value ? minFee : fee), undefined);
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/fee.service.ts:143]: Add a minObj method to Util?

File: src/subdomains/supporting/payment/services/fee.service.ts [hunk 3]
```diff
@@ -119,36 +122,61 @@ export class FeeService {
...
       this.logger.warn(`UserDiscount higher userBaseFee! UserDataId: ${userDataId}`);
-      return baseFee;
+      return { rate: baseFee.value, fixed: baseFee.fixedFee };
     }
 
-    return baseFee - discountFee;
+    return {
+      rate: baseFee.value - discountFee.rate,
+      fixed: discountFee.fixed > baseFee.fixedFee ? 0 : baseFee.fixedFee - discountFee.fixed,
+    };
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/fee.service.ts:179]: `fixed: Math.max(baseFee.fixedFee - discountFee.fixed, 0)`

File: src/subdomains/supporting/payment/services/transaction-helper.ts [hunk 4]
```diff
@@ -193,16 +194,18 @@ export class TransactionHelper implements OnModuleInit {
...
   ): Promise<TargetEstimation> {
     const price = await this.priceProviderService.getPrice(from, to);
 
     const percentFeeAmount =
-      outputAmount != null ? price.invert().convert((outputAmount * fee) / (1 - fee)) : inputAmount * fee;
-    const feeAmount = Math.max(percentFeeAmount, minFee);
+      outputAmount != null
+        ? price.invert().convert((outputAmount * fee.rate) / (1 - fee.rate))
+        : inputAmount * fee.rate;
+    const feeAmount = Math.max(percentFeeAmount + fee.fixed, minFee);
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/transaction-helper.ts:208]: `fee.fixed` is always in EUR?

File: src/subdomains/supporting/payment/services/transaction-helper.ts [hunk 5]
```diff
@@ -193,16 +194,18 @@ export class TransactionHelper implements OnModuleInit {
...
     to: Asset | Fiat,
   ): Promise<TargetEstimation> {
     const price = await this.priceProviderService.getPrice(from, to);
 
     const percentFeeAmount =
-      outputAmount != null ? price.invert().convert((outputAmount * fee) / (1 - fee)) : inputAmount * fee;
-    const feeAmount = Math.max(percentFeeAmount, minFee);
+      outputAmount != null
+        ? price.invert().convert((outputAmount * fee.rate) / (1 - fee.rate))
+        : inputAmount * fee.rate;
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/transaction-helper.ts:207]: We define `totalFee = inputAmount * fee.rate + fee.fixed` and `outputAmount = inputAmount - totalFee`.

If the user specifies an output amount, we need to take this into account.

File: src/subdomains/supporting/payment/services/transaction-helper.ts [hunk 6]
```diff
@@ -200,12 +200,16 @@ export class TransactionHelper implements OnModuleInit {
     to: Asset | Fiat,
   ): Promise<TargetEstimation> {
     const price = await this.priceProviderService.getPrice(from, to);
+    const eurSourcePrice = await this.priceProviderService.getPrice(this.eur, from);
+
+    const fixedFeeInSource = eurSourcePrice.convert(fee.fixed);
+    const fixedFeeInTarget = price.convert(eurSourcePrice.convert(fee.fixed));
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/transaction-helper.ts:206]: Move this the `convertToSource` and `convertToTarget` methods?

File: src/subdomains/supporting/payment/services/transaction-helper.ts [hunk 7]
```diff
@@ -195,21 +199,18 @@ export class TransactionHelper implements OnModuleInit {
     inputAmount: number | undefined,
     outputAmount: number | undefined,
     fee: FeeDto,
-    minFee: number,
+    { minFee: minFeeSource, fee: feeSource }: TxSpecExtended,
+    { fee: feeTarget }: TxSpecExtended,
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/transaction-helper.ts:203]: Remove fee param?

File: src/subdomains/supporting/payment/services/transaction-helper.ts [hunk 8]
```diff
@@ -245,12 +246,16 @@ export class TransactionHelper implements OnModuleInit {
       minFee: this.convert(minFee, price, from instanceof Fiat),
       minVolume: this.convert(minVolume, price, from instanceof Fiat),
       maxVolume: maxVolumeSource && this.roundMaxAmount(maxVolumeSource, from instanceof Fiat),
+      fee: {
+        rate: fee.rate && this.convert(fee.rate, price, from instanceof Fiat),
+        fixed: fee.fixed && this.convert(fee.fixed, price, from instanceof Fiat),
+      },
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/transaction-helper.ts:252]: Rates don't have a unit and cannot be converted 😉

File: src/subdomains/supporting/payment/services/transaction-helper.ts [hunk 9]
```diff
@@ -261,6 +266,10 @@ export class TransactionHelper implements OnModuleInit {
       minFee: this.convert(minFee, price, to instanceof Fiat),
       minVolume: this.convert(minVolume, price, to instanceof Fiat),
       maxVolume: maxVolumeTarget && this.roundMaxAmount(maxVolumeTarget, to instanceof Fiat),
+      fee: {
+        rate: fee.rate && this.convert(fee.rate, price, to instanceof Fiat),
+        fixed: fee.fixed && this.convert(fee.fixed, price, to instanceof Fiat),
+      },
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/transaction-helper.ts:272]: Same here

**Review by davidleomay — APPROVED** [review]

**Pull request merged**